### PR TITLE
BUG: Fix to_dict() problem when using datetime DataFrame #11247

### LINKED ISSUE #11247: DataFrame.to_dict(orient='records') datetime conversion inconsistency
My original code used:

```
data = df.to_dict(orient='records')
```

but I had to replace with a different to_dict conversion because to_dict(orient='records') doesn't properly convert numpy.datetime64 to Timestamps but rather returns numpy.datetime64 values which are hard to work with downstream.  If you include any other non-datetime fields in the dataframe the conversion to a Timestamp happens with 'records' mode, so I think this is some kind of edge-case where converters aren't being triggered when everything is a datetime.

```
import pandas as pd
df = pd.DataFrame({
        'c1': pd.to_datetime(['1-1-2015', '1-2-2015', '1-3-2015']),
        'c2': pd.to_datetime(['2-2-2015', '2-3-2015', '2-4-2015']),
    })
print('raw dataframe')
print(df)
print('\nconverted to records')
print(df.to_dict(orient='records'))
print('\nconverted to dict')
print(df.to_dict(orient='dict'))
print('\nconverted to list')
print(df.to_dict(orient='list'))
```

produces

```
raw dataframe
          c1         c2
0 2015-01-01 2015-02-02
1 2015-01-02 2015-02-03
2 2015-01-03 2015-02-04

converted to records
[{'c2': numpy.datetime64('2015-02-01T16:00:00.000000000-0800'), 'c1': numpy.datetime64('2014-12-31T16:00:00.000000000-0800')}, {'c2': numpy.datetime64('2015-02-02T16:00:00.000000000-0800'), 'c1': numpy.datetime64('2015-01-01T16:00:00.000000000-0800')}, {'c2': numpy.datetime64('2015-02-03T16:00:00.000000000-0800'), 'c1': numpy.datetime64('2015-01-02T16:00:00.000000000-0800')}]

converted to dict
{'c2': {0: Timestamp('2015-02-02 00:00:00'), 1: Timestamp('2015-02-03 00:00:00'), 2: Timestamp('2015-02-04 00:00:00')}, 'c1': {0: Timestamp('2015-01-01 00:00:00'), 1: Timestamp('2015-01-02 00:00:00'), 2: Timestamp('2015-01-03 00:00:00')}}

converted to list
{'c2': [Timestamp('2015-02-02 00:00:00'), Timestamp('2015-02-03 00:00:00'), Timestamp('2015-02-04 00:00:00')], 'c1': [Timestamp('2015-01-01 00:00:00'), Timestamp('2015-01-02 00:00:00'), Timestamp('2015-01-03 00:00:00')]}
```

### PR DESCRIPTION
**Pull request by ethanluoyc**
closes #11247

This is my first time contributing through PR so forgive me if I make any silly mistake.
The issue can be referred to in GH11247
I followed @jreback 's suggestion to box things up in `to_dict()` but I am not sure if this is what he means because I do not know exactly when `_maybe_datetime()` should be used. Initially I tried to tweak `internals.py` but it fails.

Let me know how I can improve from here. Hope I can really learn from my first PR.

### CODE REVIEW

**Comment by jreback** [comment]
see guidelines here: http://pandas.pydata.org/pandas-docs/stable/contributing.html

dont' turn on PEP changes for your editor.

**Comment by ethanluoyc** [comment]
I think I forgot to fetch to update my local master before my first commit to this PR. Somehow I messed up is it true?

**Comment by jreback** [comment]
add tests! (ideally for each one of the orients if we dont' have already).

**Comment by ethanluoyc** [comment]
which test case shall I add in? I am new to this :)

**Comment by ethanluoyc** [comment]
@jreback I have added tests:)

File: pandas/tests/test_frame.py [hunk 1]
```diff
@@ -4728,6 +4728,36 @@ def test_to_dict(self):
             for k2, v2 in compat.iteritems(v):
                 self.assertEqual(v2, recons_data[k2][k])
 
+    def test_to_dict_timestamp(self):
+        # GH11247
+        tsmp = Timestamp('20130101')
+        test_data = DataFrame({'A': [tsmp, tsmp], 'B': [tsmp, tsmp]})
```
Comment (jreback) [inline pandas/tests/test_frame.py]: can you make this a mixed frame ,e.g. add a float and a string say.

**Comment by jreback** [comment]
pls add a whatsnew note for 0.17.1 in bug fixes

**Comment by ethanluoyc** [comment]
Just asking what I should do if I merge updates locally with my branch and need to merge the 2 commits into one..



—

File: pandas/tests/test_frame.py (hunk 1 above)
Comment (ethanluoyc) [inline pandas/tests/test_frame.py]: I adjusted the tests and now it tests both mixed and single dtypes

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/89b4e5b7337606f3e1e40a3d13be863a8be6ff44

thanks!

**Pull request closed without merging**